### DISCUSSION TITLE
Seeking Advice on Enhancing Chord Name Display

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#280] [General]
Hey,

I'm currently exploring ways to enhance the visibility and aesthetics of chord names in my project. Using the Chord().GetNames() method to display all elements simultaneously seems a bit cluttered and visually unappealing. I've considered using Unicode for musical notations, but it doesn't seem to be the ideal solution.

I prefer not to develop a custom rendering engine for this purpose, nor do I intend to use a workaround involving the replacement of vector graphics with characters.

An idea I'm considering is the use of custom fonts specifically designed for chord symbols. I've found a couple of promising options [this](https://www.chordsymbols.com/) and [this](https://youtu.be/7rXOEzzMByE?si=Z-ps6IwzRjWSo_Ww&t=140), but I'm not entirely convinced they're the best fit for my needs.

I have a few questions:

1. Are you aware of any other custom fonts that might be suitable for displaying chord symbols?

2. Regarding the [Chord names table](https://melanchall.github.io/drywetmidi/articles/music-theory/Chord.html#chords-names-table), are there any specific aspects I should be cautious about? Would covering all the names in the table eliminate any potential issues?

3. Do you have any suggestions or advice on how to achieve this goal with minimal impact on latency?

Thank you for your assistance.

**Comment by melanchall** [comment]
Hi,

1. Unfortunately no. I didn't even know about such fonts :-)
2. CNT shows **all** possible suffixes for chords the library can output.

To understand better what the problem is and what are you trying to fix, can you please show your issue? Maybe a screenshot or something like that.

Thanks,
Max

**Reply by Jenix-Park** [reply]
Hey, thanks for getting back to me so fast!

You know, diving into the world of custom chord fonts was a bit new to me too. I did some digging around and learned a bunch about the tricks music notation apps use. If I were just picking a chord or two from the Chord.GetNames() output, things would be simple. But I'm aiming to display the whole shebang, so I really need to step up the game in terms of visibility and style.

So, here's what happens when I throw random note numbers into the mix. The screen ends up showing something like this:
C#m9/G, Em6/D#, C#ø/D#, D#7#5/E, D#7+5/E, C#ø7/D#, EmM7/C#, C#min9/G, D#aug7/E, Emin6/D#, C#m7(9)/G, C#m7b5/D#, C#min7(9)/G, C#m7dim5/D#, C#min7b5/D#, EminMaj7/C#, C#min7dim5/D#

As you can see, it's pretty jam-packed, and readability takes a hit. While snooping around, I realized how much musicians value the visual clarity of their sheets. This discussion [here](https://musescore.org/en/node/322579) is one example of that.

That got me thinking about ways to enhance the user experience with some snazzy chord symbols.

But after this chat, I'm feeling like maybe I need to mull it over a bit more.

Thanks a ton!

**Reply by melanchall** [reply]
Well, I see the problem now. But sorry, I have no ideas how to handle it :-(

**Reply by Jenix-Park** [reply]
Hey,

No worries at all! I really appreciate your quick responses and your willingness to help out. It's been great discussing this with you, and even though it's a tricky issue, your insights have been valuable. Thanks again for your time and support.

Take care!

**Reply by Jenix-Park** [reply]
@melanchall 
Hey,

After giving it some thought, I've come to the conclusion that the simplest solution is to just replace all '#' characters with the Unicode '♯' and the 'b's with '♭'. Implementing this change on our end is nothing, but having an additional option directly in your plugin would really be great in terms of both performance and ease of use.

What I'm thinking is, maybe the Chord().GetNames() method could accept a boolean argument. This would let us decide whether to use the Unicode characters '♯' and '♭' in the output. If this parameter is set to False or left out, it sticks with the current style. But if we pass in True, it switches over to the new Unicode style.

Just tossing out an idea here.

Thanks a bunch!

**Reply by melanchall** [reply]
Hi @Jenix-Park,

Looks interesting. I think something like `GetNames(ChordNameHint.UseStyledNames)` would be nice. A `bool` parameter is not flexible if one day we'll decide to add another one option. "Hint" way provide us a space for extending true/false options. For example, `GetNames(ChordNameHint.UseStyledNames | ChordNameHint.ReturnMostRelevant)`.

What do you think?

**Reply by Jenix-Park** [reply]
Absolutely, I'm totally on board with the GetNames(ChordNameHint.UseStyledNames) approach. It's a flexible and forward-thinking solution. Thanks a lot for coming up with this – it's perfect!

Cheers!

**Reply by Jenix-Park** [reply]
@melanchall
Hey,

I think we might not need to make any changes after all. I tested the [custom font](https://www.youtube.com/watch?v=7rXOEzzMByE&t=140s) I mentioned earlier, and it turns out the current style actually works quite well with it. You might want to give it a try yourself; it's available for free download – just enter $0.

While we still need to tweak the output text for more complex chords, simpler ones like D#b5 look pretty sharp as is, with no modifications needed.

Thanks so much for your time and interest in this matter!

**Reply by melanchall** [reply]
Hi @Jenix-Park,

Well, I'm going to put the feature on hold then :-) Thanks for musical fonts information!

**Reply by Jenix-Park** [reply]
@melanchall 
Hey again!

After testing for a while, I've been thinking about the issue with crowded chords we talked about earlier.
I realized I might not necessarily need special fonts or musical symbols to solve it.

You know, the '#'s, '/', and 'b' characters in the ASCII code have been widely used and understood in music sheets for years.
While using dedicated musical symbols might make things look cooler, we can also make chords easier to read by simply using fewer of them.

At first, I threw all the chord names at users to cover all types of musicians.
But what if we gave users some choices like:

GetNames(ChordNameHint.JazzStyle)
GetNames(ChordNameHint.ClassicalStyle)

This way, they can pick how the chords look – a bit jazzy or with a classical touch.
What do you think about this idea?

**Reply by melanchall** [reply]
Hi @Jenix-Park,

It's an interesting idea. But unfortunately I don't know what kind of chord notation is jazz or classical one :-(

**Reply by Jenix-Park** [reply]
@melanchall 
No worries!
I'm not a musician myself, but I figured we could tap into ChatGPT for insights on chord notations used by both jazz and classical musicians.

I took a peek at the 'Chords names table.' Initially, I thought you might already have some jazz-style symbols like Δ and +, and a little tweaking with some additional if conditions could do the trick. Turns out, I was mistaken, and it seems like a bit more work is needed.

Thanks for looking into it!
